### DISCUSSION TITLE
Single file support

### DISCUSSION THREAD

**Discussion by ceed0** [joaotavora/eglot#1086] [General]
Eglot always sets current directory as a project, instead of relaying on project.el. This prevents using LSP in a single file mode.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_workspaceFolders

> The [workspace/workspaceFolders](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_workspaceFolders) request is sent from the server to the client to fetch the current open list of workspace folders. Returns `null` in the response if only a single file is open in the tool. Returns an empty array if a workspace is open but no folders are configured.

**Comment by nemethf** [comment]
Eglot does rely on project.el, which does not directly support a single-file project as `project-root` always returns something.

**Reply by ceed0** [reply]
Yeah, but it seems to rely on a current dir as a fallback. I think it shouldn't - if project.el doesn't detect a project, eglot should assume there's none. If a user wants to use current dir as a project there is already a prompt in project.el to set a directory as a transient project, and you can even define your own functions to automatically find projects with `project-find-functions` variable.

Here https://github.com/joaotavora/eglot/blob/master/eglot.el#L1005 it fallbacks to using `default-directory` as a transient project.

**Reply by joaotavora** [reply]
"Current dir as fallback" is a already a `project.el` mechanism, as you note.  Eglot merely streamlines it. Also there's nothing preventing you from using `project-find-functions` or from integrating EDE, or anything as long as `project.el` recognizes it.  

Why don't you explain what you want to do, but can't?  What is the benefit to the user (and under which conditions and in which language server(s)) that the response to `workspace/workspaceFolders` is a `:null`, instead of an array with a single element?  

Once we have this answer and if there is some reasonable advantage to gain from that, then maybe something can be added in `project.el` which Eglot can consult to tweak the value to `:null` in some cases.

**Reply by joaotavora** [reply]
> if project.el doesn't detect a project, eglot should assume there's none.

I'm sorry to be blunt here but this idea would single-handedly break one of the best Eglot features, which is that it works fine for a large number of servers in the single-file case already:  creating a single `.c` file in some new dir, typing `M-x eglot`  then working on it, then adding a `.h` file as a dependency just works :tm: .  No knowledge of projects ever required for this.

**Reply by ceed0** [reply]
> What is the benefit to the user

Benefit is that it's more predictable and won't slow things down in case I don't create a new directory for a single file, but use it as is. 

Consider that a file can be placed right in the root of a home dir. In that case with [pylsp](https://github.com/python-lsp/python-lsp-server), for example, if you try to find xref it will just time out every time. And depending on how a server was implemented I think it could lead to an unwanted code execution which isn't great either.

If you want to streamline this, it would be better to add an option to project.el, instead of forcing this with eglot.

**Reply by joaotavora** [reply]
So you put a bunch of unrelated python files in your home dir, i.e. `~`? I don't think that's a common use case, at all. A much more common use case, IMO -- not to mention one that shall not be broken since it's been in place for a number of years -- is to use Eglot with a single file in a single directory, then add more files, version control, sub-directories, `.dir-locals`, etc, etc

But you can ask `project.el`'s maintainer (@dgutov) to help you define single-file projects comprised only from such files as you have in your home directory.  Once that is in place Eglot can consult some hypothetical `project-single-file-p` predicate and accordingly inform the LSP server that there is only a single file being handled.

Alternatively, for servicing your use case, you can try use Elisp to add a a method to the  `eglot-workspace-folders` generic function.  See the Elisp manual entry on `defmethod` and generic functions.

**Reply by dgutov** [reply]
> project-root always returns something

`project-current`, however, can return nil (if called without arguments). At which point you are free to use different grouping strategies, such a prompting for a dir, or using the current file, etc.

Changing the project api to make "single file project" possible is something along the lines of what people had asked for in the past, but I'm not sure it will be useful enough in general. Most of the time people want the root directory, and if it's not guaranteed to be there, well...

**Reply by ceed0** [reply]
> A much more common use case, IMO -- not to mention one that shall not be broken since it's been in place for a number of years -- is to use Eglot with a single file in a single directory, then add more files, version control, sub-directories, .dir-locals, etc, etc

For this use-case there would be almost no difference, because vcs would be used anyway or a function to detect `.dir-locals` can be easily added to `project-find-functions` etc.

Meanwhile this can break other uses, as I've shown. Also, consider that lsp can be used as a tool for more than a code (as a [grammar checker](https://github.com/languagetool-language-server/languagetool-languageserver) for example) and it can be used in different environments.

**Reply by joaotavora** [reply]
> For this use-case there would be almost no difference, because vcs would be used anyway or a function to detect .dir-locals can be easily added to project-find-functions etc.

We're miscommunicating.  Currently, `M-x eglot` works without any dir-locals, without .git, without any Elisp configuration.  Just with a single file.  

Literally, this works:
```
emacs -Q -l eglot /tmp/newdir/newthingy.c -f save-buffer -f eglot
```

I don't want this to stop working.  This is one of the best Eglot features, I'm not going to break it.  So "almost no difference", whatever that "almost" encompasses for you, is a non-starter for me.

> Meanwhile this can break other uses, as I've shown. Also, consider that lsp can be used as a tool for more than a code (as a [grammar checker](https://github.com/languagetool-language-server/languagetool-languageserver) for example) and it can be used in different environments.

That's great, but I don't see the relation.  The only use problematic case you've shown so far is the case of storing single-file Python programs in the root of a very large hierarchy, i.e. your $HOME folder.  I don't consider that case to be common at all (or even a good idea, but that's none of my business).  I've given you advice as to how to to proceed to eventually have that case supported in Eglot,  realized in tandem with some support from `project.el`.

PS: thanks for the LSP link as grammar checker, this may come in handy for a forthcoming feature.

**Reply by ceed0** [reply]
All I'm trying to say is that there are already plenty of options for users who want to use current dir as a project. And you can prompt users or it would be easy to add an option to project.el for eglot's current behavior.

I see no reason why managing single files should be pushed to project.el, since it exists for managing more-than-one-file projects, adding predicates and such for some single-file-project would be just adding more cruft. While eglot shouldn't take it on itself to define project directories - it's a tool to interact with lsp, and lsp spec defines how to manage single files already.

**Comment by joaotavora** [comment]
I understand, but to be clear here, your opinion as to how Eglot should function is not nearly as useful as answers to these simple questions:

1. Is your suggestion enabling a new use case that isn't already covered? You have brought forth only one, a vali one, but reasonably odd. I've also suggested ways to work around your problem.

2. Does implementing your suggestion break and existing behavior for other users who are not you? It seems it does.

So, i can  only consider suggestions that answer 1 and 2 with  "yes" and "no", respectively.

**Reply by ceed0** [reply]
> Is your suggestion enabling a new use case that isn't already covered? You have brought forth only one, a vali one, but reasonably odd. I've also suggested ways to work around your problem.

It's not covered. You proposed to add some code manage single files in project.el. But that would be a cruft, because emacs already works well with single files, adding project.el to the mix is pointless. Alternative that you also proposed, would be to hack into eglot guts, which is strange, considering that this would be a workaround for a bug. And it is a bug, not a feature, even if it's somehow useful, because it doesn't comply with the spec and language servers would expect different behaviour. That it works in the most popular use-case is just a coincidence.

If you want a less "odd" example consider [bash lsp](https://github.com/bash-lsp/bash-language-server) everyone's bashrc or equivalent is placed right in `$HOME`, now, if you run lsp on bashrc, as you might've guessed, it won't go well - right from the start, bash-lsp will try to crawl everything inside `$HOME`, and then eventually will just die. But really this could be anything beside well defined scenarios with projects in vcs or "one file in a new directory".

> Does implementing your suggestion break and existing behavior for other users who are not you? It seems it does.

You mean "zero-config from cli" one? How many users really use emacs like that and without any config at all? I can't imagine anyone beside someone who just misclicked on emacs' desktop icon or something. I hope you don't honestly trying to protect this as a legit use-case.

Other than that, my suggestion is to make current behaviour optional, and move the option from eglot to the project.el instead, where it can be used by others as well, not only by eglot users. Or leave it in eglot, maybe even leave it as a default (I doubt project.el would accept such option as a default because that not really what users would expect), but instead add an option to use eglot in a single-file mode, that way you don't have to change default behaviour at all (even if in my opinion it would be a wrong way to go).

Or, instead of an option, eglot could behave like project.el already does - if I run `project-find-regexp` or some other project.el function outside of a detected project it would give me a prompt first, there I can either choose a project from history, or choose a dir as a transient project. Expanding on this, eglot could run in a single-file mode if user cancelled a selection. Also, for (almost) zero-conig experience eglot function could accept an argument to revert to current behaviour. So example above would turn into something like:

`emacs -Q -l eglot /tmp/newdir/newthingy.c -f save-buffer --eval "(eglot t)"`

**Reply by joaotavora** [reply]
> bash-lsp will try to crawl everything inside $HOME, and then eventually will just die. 

If it behaves like it says you do, then seems more like a problem of bash-lsp not recognizing the exceptional nature of `~/.bashrc`, which is again odd since `/usr/bin/bash` itself clearly recognizes it.

> You mean "zero-config from cli" one? How many users really use emacs like that and without any config at all? I can't imagine anyone beside someone who just misclicked on emacs' desktop icon or something. I hope you don't honestly trying to protect this as a legit use-case.

You're completely missing the point.  I just gave that example because it fits in a single line and I don't have all day to address random internet opinions.  You could just as well start Emacs in whatever clikcy in Windoze 3000 way and open that file with a super-fancy file browser and type `M-x eglot`: it's exactly the same.

> Other than that, my suggestion is to make current behaviour optional, and move the option from eglot to the project.el instead, where it can be used by others as well, not only by eglot users

I don't understand your suggestion anymore.  It seems you are proposing more or less the same I am.  But I'm not sure anymore.  Tell you what, since Eglot is moving to Emacs very soon, you'll be able to address this with a patch to Emacs.git and then we can see what you mean and what is broken and what is not.

> lternative that you also proposed, would be to hack into eglot guts, which is strange, considering that this would be a workaround for a bug. 

Adding methods to generic functions is part of Eglot's API, and a way to interact with programs that's more than 40 years old. That you think it's strange and a hack is totally irrelevant to me: it was my honest suggestion to address your problem.

**Reply by ceed0** [reply]
> I don't understand your suggestion anymore.

1. In eglot if `project-current` returns nil, instead of forcing use of current directory, use single file
2. Add option to project.el that if something calls `project-current` and there's no project return `default-directory` without a prompt

Best case scenario of course. Instead of that, everything can stay the same and an option can be added in eglot to enable single file mode, or I can just hack it in, or whatever - I got that it's pointless to try and convince you, so I'll just stop.

**Comment by joaotavora** [comment]
> 1. In eglot if project-current returns nil, instead of forcing use of current directory, use single file

So you insist that Eglot's current behaviour of 4+ years  be changed, disturbing the workflows of an unknown number of users, in which I count myself.  Well, I don't know how else to say this.  **No. It won't happen.**

For the record, my 2 suggestions to encompass the "forced single-file" case in Eglot:

Suggestion 1:

* Add some way to project.el do define a single-file-never-will-have-two-files-project, if that way doesn't exist already
* Add a predicate `project-is-single-file-will-never-have-two-files-p`.  
* Have Eglot consult this predicate in the default method of the `eglot-workspace-directories` generic function.

Suggestion 2:

* The user defines an `:around` method added to the `eglot-workspace-directories` generic function that somehow figures out if the project is single-file-will-never-have-two-files-p and returns nil in such a case.

None of these suggestions break the current behavior.

**Comment by ceed0** [comment]
So even a slight change in a workflow, like I proposed earlier (quoted below), is a hard no?

> eglot could behave like project.el already does - if I run `project-find-regexp` or some other project.el function outside of a detected project it would give me a prompt first, there I can either choose a project from history, or choose a dir as a transient project. Expanding on this, eglot could run in a single-file mode if user cancelled the selection

This could also be topped with an option to revert to the old behavior maybe even in project.el itself, granted, I'm not sure how this could be implemented, that's just an idea.

> Suggestion 1

I really don't understand what the point for such predicate would be, `project-current` already returns nil for single files, which already can be used as a predicate to add an option for a single-file use case, even if you don't want this option to be turned on by default.

But if you don't want to see any options either I personally can go down the second route, even if for me, and whoever else would want this, it would be far from no-config.

**Reply by joaotavora** [reply]
> So even a slight change in a workflow, like I proposed earlier (quoted below), is a hard no?

As I explained already: It's not a "slight" change by any measure.  It's a change that requires users that have so far been using Eglot with projects in a certain way to go out of their way to configure variables, initialize version control systems, etc.

What's more, even if that were somehow tolerable -- which it's not -- I'm not convinced that the single-file-forever case is a more common case than the single-file-then-more-files case.

> I really don't understand what the point for such predicate would be, project-current already returns nil for single files, which already can be used as a predicate to add an option for a single-file use case, even if you don't want this option to be turned on by default.

Let me explain.  Eglot needs projects, `project.el` projects. That's an invariant.  The simplest representations of projects in `project.el` is a root directory as @nemethf already explained.  But `project.el` is flexible enough to allow you to represent projects with certain constraints, for example with [EDE](https://www.gnu.org/software/emacs/manual/html_mono/ede.html#Creating-a-project), or with any number of ways using its API.  One of those constraints would be "this project has this single file".  If `project.el` exposed a predicate that flagged if such a constraint is active, then Eglot could consult that predicate to tell the LSP server "there are no 'workspaces' for this LSP session".

> it would be far from no-config.

How "far" remains to be seen.  But basically yes: Eglot's "no config case" is aimed towards projects in directories. It works well with single-file except, presumably, in situations such as yours which are, arguably, the minority.

**Reply by dgutov** [reply]
Eglot could grow a new user option that would default to the current behavior.

**Reply by ceed0** [reply]
I'm sorry, i guess you're just joking now firs you wrote "go out of their way to configure variables", and then "How "far" remains to be seen" in the context of writing a new method for an internal function of eglot, which would require knowledge of elisp and eglot far greater, than "going of my way" to configure a variable, even if result would look small in size.

**Reply by joaotavora** [reply]
> Eglot could grow a new user option that would default to the current behavior.

It's a property of the project whether the project is single-file-only or not. It makes no sense to have Eglot record that fact as a global user option, as multiple projects may exist, some single-file-only others just single-file-for-now. 

But I think Eglot can happily consult that fact about a given project using `project.el`'s API, if/when such an API endpoint exists. 

Mind that I'm not saying this is an enormous shortcoming of `project.el`.  I personally think it's mostly fine the way it is and this use case is easily solved with an Elisp user snippet. But it is a shortcoming there, not here.

> I'm sorry, i guess you're just joking now firs you wrote "go out of their way to configure variables", and then "How "far" remains to be seen" in the context of writing a new method for an internal function of eglot, which would require knowledge of elisp and eglot far greater, than "going of my way" to configure a variable, even if result would look small in size.

How can it be so hard to understand that users upgrading to a hypothetical new version of Eglot with your proposed default behaviour would now be unpleasantly surpsied to find they have to configure a new variable to recover behavior that has been default for a number of years and Eglot versions?   

Using Elisp snippets to address your own particular problems is perfectly standard practice in Emacs.  We can and we do publish such snippets in the manual, when the need is pervasive enough among many users.

**Reply by ceed0** [reply]
> How can it be so hard to understand that users upgrading to a hypothetical new version of Eglot with your proposed default behaviour would now be unpleasantly surpsied to find they have to configure a new variable to recover behavior that has been default for a number of years and Eglot versions?

That's what versioning is for, no? And in light of eglot moving to mainline, I think such a change could be done. Especially since this would be fixing eglot's implementation of lsp protocol to be in line with the spec.

But indeed I don't see an easy way for eglot to find out when new project was defined with project.el beside maybe running a timer on `project-current` function?

> Using Elisp snippets to address your own particular problems is perfectly standard practice in Emacs.

Yeah, I know, just found it funny.

**Reply by joaotavora** [reply]
> That's what versioning is for, no? 

No. In my book, versioning is _not_ for throwing backward compatibility out of window gratuitously.

> Eglot's implementation of lsp protocol to be in line with the spec.

There nothing out of line with the spec here.  Emacs project infrastructure doesn't support single-file-only projects.  It supports projects that are composed of a single file temporarily but not exclusively.  All Emacs's projects have a `project-root`, and a root is a directory.  If/\when that situation changes, Eglot will pass that informatino to the LSP server.  Until then it'll work like it has worked basically forever.

**Reply by dgutov** [reply]
> I personally think it's mostly fine the way it is and this use case is easily solved with an Elisp user snippet. But it is a shortcoming there, not here.

It might only be declared a shortcoming there if other clients of the api determine the need to ask the same questions. But this is the first time I've head about "single file projects".

**Reply by ceed0** [reply]
> There nothing out of line with the spec here.

That's where we have to agree to disagree, I guess. Spec doesn't require a "project" or some equivalent to exist, if a single file is opened `workspace/workspaceFolders` should return `null`, and even if project.el doesn't support this eglot could.

**Reply by joaotavora** [reply]
> It might only be declared a shortcoming there if other clients of the api determine the need to ask the same questions. But this is the first time I've head about "single file projects".

Well I have lots and lots of single file projects, most of my projects do start with a single file. I always store them in directories, so that if the project eventually grows more files, those files will be part of the project.  

Here, I think what's missing from `project.el` or thereabouts is to be able to define a project with a fixed set of files, say A, B, and C so that adding another file D in any of A, B, and C's directories does _not_ automatically add D to the project's files.  Is there a way to do this now within Emacs's project infrastructure? If so, Is there a way to ask of a given project returned by `project-current` whether it has this special property?

**Reply by dgutov** [reply]
> Well I have lots and lots of single file projects, most of my projects do start with a single file. I always store them in directories, so that if the project eventually grows more files, those files will be part of the project.

What would be an automated way to detect that a project is single-file?

> Is there a way to do this now within Emacs's project infrastructure? If so, Is there a way to ask of a given project returned by project-current whether it has this special property?

You can call `project-files`, but `project-root` returning a valid directory is currently mandatory. I suppose it's up to you to detect that `project-files` returns just one file. But having such a project returned is non-trivial for the user.

**Reply by joaotavora** [reply]
> I suppose it's up to you to detect that project-files returns just one file. 

At first, I also thought that was enough.  But its not.  The client of this information needs to know that if new files are added to the project's `project-root`, they will _not_ automatically be added to `project-files` automatically.
     
> What would be an automated way to detect that a project is single-file?

I don't know. By "automated" I take it you mean a way in which the user doesn't explicitly designate a project as single-file  (or fixed files). If so, I don't think one exists, at least not one backward-compatible with the current way `project.el` works (which I think is the correct way).
  
Maybe a buffer-local variable could help -- `project-current` would return a special object when invoked in a buffer with this cookie -- but that sounds hacky because I don't know if it's a nice idea to assume that project objects are always returned by `project-current`. So it would have to be a database in a user Elisp variable or a some value in a supporting file, like .dir-locals.

That said, my opinion is that none of this is priority.

I think the OPs case should be solved by rethinking his/her  organization of files or by using something like:
```
(defmethod eglot-workspace-folders :around (server)
  (unless (equal (project-files (eglot--project server)) '("~/.bashrc"))
    (call-next-method)))
```
Or something like that.  Eventually this would require Eglot to expose `eglot--project` as `eglot-project`, but that's fine.

**Reply by dgutov** [reply]
> if new files are added to the project's project-root, they will not automatically be added to project-files automatically

Why not call `project-files` again and compare?

> Maybe a buffer-local variable could help -- project-current would return a special object when invoked in a buffer with this cookie -- but that sounds hacky because I don't know if it's a nice idea to assume that project objects are always returned by project-current. So it would have to be a database in a user Elisp variable or a some value in a supporting file, like .dir-locals.

If some suitable project was defined, I suppose the relevant buffers could have an `eval: (setq project-find-functions '(one-element))` somewhere. But "suitable" is still under question, as I understand.

> That said, my opinion is that none of this is priority.

Very good.

**Reply by joaotavora** [reply]
> Why not call project-files again and compare?

Because the goal here is to inform the language server _upfront_that it needn't concern itself with any other files other than the ones mentioned in the `textDocument/didOpen` notifications.  So there is no moment in time where this repeated call would make any sense. By default, Eglot informs not only informs the server that it opened file A but that that file lives under one of multiple project roots ("workspace folders" in LSP parlance).  The server may then start searching/indexing those roots for more information it if wants to.

> If some suitable project was defined, I suppose the relevant buffers could have an eval: (setq project-find-functions '(one-element)) somewhere. But "suitable" is still under question, as I understand.

Something like that, yes.  Though mind you: it's not just "single file" it would be more like "fixed files", of which "single file" is a special case. The server is probably OK with being passed other files in `didOpen` notifications, the OP here just wants for a way so that Eglot can tell the server that it needn't concern itself with any other files other than the ones mentioned in `didOpen`.

**Reply by ceed0** [reply]
> I think the OPs case should be solved by rethinking his/her organization of files or by using something like:
> 
> ```
> (defmethod eglot-workspace-folders :around (server)
>   (unless (equal (project-files (eglot--project server)) '("~/.bashrc"))
>     (call-next-method)))
> ```

Scripts can be placed in random places, have random names, I can't write them all down in my config. Current eglot's behavior prevents me from starting it with `python-mode-hook` and the like, because at best it wild lead to a subpar experience, and at worst to some nasty code execution.

And your assumption that my file organization is somehow wrong or a minority case is completely baseless.

**Reply by joaotavora** [reply]
> some nasty code execution.

Eglot doesn't execute anything but its own Lisp code.

> Scripts can be placed in random places, have random names, I can't write them all down in my config

I've just suggesting a general approach. By harnessing the power of programming, you can devise some way to designate catch these scripts, perhaps by checking if they are located in the home directory. 

> And your assumption that my file organization is somehow wrong or a minority case is completely baseless.

It is based on reports that you place python and other programs directly in your home directory. 

Also, I'm done with trying to help you. Have a nice day.

**Comment by MassimoLauria** [comment]
Sorry to resurrect the discussion, but I have a similar problem and I found myself trying to get something out of it. I sometime write throwaway scripts and programs, usually from $HOME, to file them elsewhere if they become something I want to keep around. When `eglot` kicks in, Emacs hangs for quite a bit. I guess that's because `find` scans the whole home, and `eglot` send filenames to the `LSP`. 

My current solution is to use this function instead of `eglot-ensure`
```emacs-lisp

(defun mxl/eglot-ensure-not-home ()
  "Activate eglot only if not in $HOME"
  (interactive)
  (if (string=
       (expand-file-name default-directory)
       (expand-file-name "~/"))
      (message "Eglot forbidden in ~/")
    (eglot-ensure)))
```

I would rather have `eglot` active in single-file mode whenever it is active on a buffer in the home directory. I don't know much emacs-lisp and I tried to add this snippet that I interpolated from the previous discussion. But it did not work.

```emacs-lisp 
(cl-defmethod eglot-workspace-folders :around (server)
           (unless (string= (expand-file-name default-directory)
                            (expand-file-name "~/"))
             (call-next-method)))
```
